### PR TITLE
interfaces: API additions for interface hooks

### PR DESCRIPTION
**Pull request by stolowski**
Changes to the interface of interfaces: extended AddConnectedPlug/Slot methods to take slot/plug attributes maps. Added stubs for ValidatePlug/Slot methods which will be called to validate dynamic attributes coming from interface hooks (this is implemented in PR #3120).

The API will be used in a followup PR.

### CODE REVIEW

File: interfaces/builtin/bluez.go [hunk 1]
```diff
@@ -225,3 +225,13 @@ func (iface *BluezInterface) AutoConnect(*interfaces.Plug, *interfaces.Slot) boo
 	// allow what declarations allowed
 	return true
 }
+
+func (iface *BluezInterface) ValidatePlug(plug *interfaces.Plug, attrs map[string]interface{}) error {
```
Comment (zyga) [inline interfaces/builtin/bluez.go]: I kind of wish we did the same trick we have elsewhere with interfaces. We could have a function that validates a plug that checks if the interface supports validation. This would cut the amount of boilerplate code *significantly* as it could do simple stuff like checking if interfaces match etc etc (sanitize could be done the same way). What do you think?

File: interfaces/builtin/unity8_pim_common.go [hunk 2]
```diff
@@ -131,21 +127,13 @@ func (iface *unity8PimCommonInterface) AppArmorConnectedPlug(spec *apparmor.Spec
 	return nil
 }
 
-func (iface *unity8PimCommonInterface) ConnectedPlugSnippet(plug *interfaces.Plug, slot *interfaces.Slot, securitySystem interfaces.SecuritySystem) ([]byte, error) {
```
Comment (zyga) [inline interfaces/builtin/unity8_pim_common.go]: I'd leave those out for a separate PR. (as a habit)

File: interfaces/builtin/bluez.go (hunk 1 above)
Comment (stolowski) [inline interfaces/builtin/bluez.go]: That's a very nice idea indeed. I can do it (in a followup)

Comment (stolowski) [inline interfaces/builtin/bluez.go]: @zyga Ok, I made ValidatePlug/Slot optional in the PR #3120 as this is where it's used (also improved a test there to make sure it's called when it should). ValidatePlug/Slot is removed from base Interface in core.go completely and I've removed it from all interfaces which have no attributes (so there is no potential use for Validate* to exists atm); I've left empty stubs for Validate* in interfaces which can potentially make use of attributes from hooks.

**Comment by zyga** [comment]
Can we please...

 - land https://github.com/snapcore/snapd/pull/3284
 - merge master into this branch
 - add extra checks in `interfaces/builtin/all_test.go` so that no legacy methods are used
 - see tests go green

... before landing this branch?
